src/porepy/viz/diagnostics_mixin.py

```diff
@@ -0,0 +1,383 @@
+                        name=eq_name,
+                        dim=grid.dim,
+                        grid_id=grid.id,
+                        is_interface=is_interface,
+                    )
+                    equation_indices[grid_equation_name] = grid_dof_indices
+                block_indices += dof_indices.size
+
+        elif grouping == "subdomains":
+            for eq_name, dof_indices in assembled_equation_indices.items():
+                equation_indices[eq_name] = dof_indices
+
+        return equation_indices
+
+    def _variable_indices(
+        self, grouping: Optional[Literal["subdomains"]] = None
+    ) -> dict[str, np.ndarray]:
+        """Collects the indices of variables presented in the Jacobian matrix as
+        columns.
+
+        Returns:
+            A dictionary with keys - printed names of the variables, values - indices of
+            this variables in the global Jacobian matrix.
+        """
+        variable_indices = {}
+        assert grouping in (None, "subdomains"), f"Unknown grouping: {grouping}"
+
+        if grouping is None:
+            for variable in self.equation_system.variables:
+                grid_variable_name = _format_variable_name_with_grid(variable)
+                dofs = self.equation_system.dofs_of([variable])
+                variable_indices[grid_variable_name] = dofs
+
+        elif grouping == "subdomains":
+            previous_variable_name = None
+            for variable in self.equation_system.variables:
+                if previous_variable_name != variable.name:
+                    dofs = self.equation_system.dofs_of(
+                        [self.equation_system.md_variable(variable.name)]
+                    )
+                    variable_indices[variable.name] = dofs
+                    previous_variable_name = variable.name
+
+        return variable_indices
+
+    def _collect_submatrices(
+        self,
+        mat: spmatrix,
+        equation_indices: dict[str, np.ndarray],
+        variable_indices: dict[str, np.ndarray],
+    ) -> None:
+        """Slices the Jacobian matrix into block based on provided equations and
+        variables.
+
+        Args:
+            mat: The full Jacobian matrix.
+            equation_indices: The dictionary with keys - written names of equations and
+                values - indices of these equations in the Jacobian. (Rows index).
+            variable_indices: The dictionary with keys - written names of variables and
+                values - indices of these variables in the Jacobian (Columns index).
+        """
+        submatrices = {}
+
+        for i, ind_row in enumerate(equation_indices.values()):
+            for j, ind_col in enumerate(variable_indices.values()):
+                submatrix_indices = np.meshgrid(
+                    ind_row, ind_col, sparse=True, indexing="ij"
+                )
+                submatrices[i, j] = mat[tuple(submatrix_indices)]
+
+        self.submatrices = submatrices
```

Comment:
Is there any reason not to return this, rather than storing it? I am afraid this will effectively lead to a double storage of the matrix.